# Smart Logistics Data Cleaning  
* Rahim Manji

In [1]:
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt

In [3]:

data = pd.read_csv('../data/smart_logistics_dataset.csv')

# 1.Data Cleaning & Prep

In [61]:
data.head()

,Timestamp,Asset_ID,Latitude,Longitude,Inventory_Level,Shipment_Status,Temperature,Humidity,Traffic_Status,Waiting_Time,User_Transaction_Amount,User_Purchase_Frequency,Logistics_Delay_Reason,Asset_Utilization,Demand_Forecast,Logistics_Delay
0,2024-03-20 00:11:14,Truck_7,-65.7383,11.2497,390,Delayed,27.0,67.8,Detour,38,320,4,NaN,60.1,285,1
1,2024-10-30 07:53:51,Truck_6,22.2748,-131.7086,491,In Transit,22.5,54.3,Heavy,16,439,7,Weather,80.9,174,1
2,2024-07-29 18:42:48,Truck_10,54.9232,79.5455,190,In Transit,25.2,62.2,Detour,34,355,3,NaN,99.2,260,0
3,2024-10-28 00:50:54,Truck_9,42.3900,-1.4788,330,Delivered,25.4,52.3,Heavy,37,227,5,Traffic,97.4,160,1
4,2024-09-27 15:52:58,Truck_7,-65.8477,47.9468,480,Delayed,20.5,57.2,Clear,56,197,6,NaN,71.6,270,1


In [62]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Timestamp                1000 non-null   object 
 1   Asset_ID                 1000 non-null   object 
 2   Latitude                 1000 non-null   float64
 3   Longitude                1000 non-null   float64
 4   Inventory_Level          1000 non-null   int64  
 5   Shipment_Status          1000 non-null   object 
 6   Temperature              1000 non-null   float64
 7   Humidity                 1000 non-null   float64
 8   Traffic_Status           1000 non-null   object 
 9   Waiting_Time             1000 non-null   int64  
 10  User_Transaction_Amount  1000 non-null   int64  
 11  User_Purchase_Frequency  1000 non-null   int64  
 12  Logistics_Delay_Reason   737 non-null    object 
 13  Asset_Utilization        1000 non-null   float64
 14  Demand_Forecast          

In [63]:
data.shape

(1000, 16)

In [64]:
data.describe(include='all').T

,count,unique,top,freq,mean,std,min,25%,50%,75%,max
Timestamp,1000,1000,2024-03-20 00:11:14,1,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Asset_ID,1000,10,Truck_8,109,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Latitude,1000.0,NaN,NaN,NaN,-1.360093,51.997183,-89.7915,-46.167975,-4.50315,44.5028,89.8701
Longitude,1000.0,NaN,NaN,NaN,0.837049,104.843618,-179.8202,-88.448075,0.6783,88.15645,179.9237
Inventory_Level,1000.0,NaN,NaN,NaN,297.915,113.554773,100.0,201.0,299.0,399.0,500.0
Shipment_Status,1000,3,Delayed,350,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Temperature,1000.0,NaN,NaN,NaN,23.8939,3.322178,18.0,21.2,23.8,26.6,30.0
Humidity,1000.0,NaN,NaN,NaN,65.0422,8.753765,50.0,57.2,65.2,72.4,80.0
Traffic_Status,1000,3,Detour,345,NaN,NaN,NaN,NaN,NaN,NaN,NaN
Waiting_Time,1000.0,NaN,NaN,NaN,35.062,14.477768,10.0,23.0,35.0,49.0,60.0


# Data Quality Assessment

- Before cleaning, we assess:  missing values, duplicate records, data types, categorical consistency, and numerical ranges

In [65]:
#missing values 

In [66]:
missing = data.isnull().sum()

missing

Timestamp                    0
Asset_ID                     0
Latitude                     0
Longitude                    0
Inventory_Level              0
Shipment_Status              0
Temperature                  0
Humidity                     0
Traffic_Status               0
Waiting_Time                 0
User_Transaction_Amount      0
User_Purchase_Frequency      0
Logistics_Delay_Reason     263
Asset_Utilization            0
Demand_Forecast              0
Logistics_Delay              0
dtype: int64

In [67]:
# Missing value percentages

missing_summary = pd.DataFrame({
    'Missing_Count': data.isnull().sum(),
    'Missing_Percentage': (data.isnull().sum() / len(data) * 100).round(2)
})

missing_summary

,Missing_Count,Missing_Percentage
Timestamp,0,0.0
Asset_ID,0,0.0
Latitude,0,0.0
Longitude,0,0.0
Inventory_Level,0,0.0
Shipment_Status,0,0.0
Temperature,0,0.0
Humidity,0,0.0
Traffic_Status,0,0.0
Waiting_Time,0,0.0


In [68]:
# Check duplicates
data.duplicated().sum()

0

- Investigate the 263 missing values in Logistics_Delay_Reason


In [69]:
# Display rows where  Logistics_Delay_Reason  is missing 


In [70]:
'''
Variables to consider for missing Logistics_Delay_Reason values :

- Shipment_Status
- Traffic_Status
- Logistics_Delay
- Logistics_Delay_Reason
'''

'\nVariables to consider for missing Logistics_Delay_Reason values :\n\n- Shipment_Status\n- Traffic_Status\n- Logistics_Delay\n- Logistics_Delay_Reason\n'

In [71]:
data[data['Logistics_Delay_Reason'].isnull()].head(10)

,Timestamp,Asset_ID,Latitude,Longitude,Inventory_Level,Shipment_Status,Temperature,Humidity,Traffic_Status,Waiting_Time,User_Transaction_Amount,User_Purchase_Frequency,Logistics_Delay_Reason,Asset_Utilization,Demand_Forecast,Logistics_Delay
0,2024-03-20 00:11:14,Truck_7,-65.7383,11.2497,390,Delayed,27.0,67.8,Detour,38,320,4,NaN,60.1,285,1
2,2024-07-29 18:42:48,Truck_10,54.9232,79.5455,190,In Transit,25.2,62.2,Detour,34,355,3,NaN,99.2,260,0
4,2024-09-27 15:52:58,Truck_7,-65.8477,47.9468,480,Delayed,20.5,57.2,Clear,56,197,6,NaN,71.6,270,1
5,2024-09-17 06:02:15,Truck_7,73.3312,46.5831,118,In Transit,24.3,61.8,Clear,56,258,10,NaN,66.8,189,0
6,2024-02-05 08:38:56,Truck_4,27.9307,147.5317,480,Delayed,20.7,75.4,Clear,32,263,3,NaN,73.3,198,1
8,2024-01-18 08:27:04,Truck_9,-83.6737,-31.7125,245,Delivered,26.4,77.2,Clear,14,183,2,NaN,69.6,206,0
10,2024-01-09 23:19:43,Truck_9,82.9633,-154.5274,319,Delayed,22.0,53.5,Clear,28,360,4,NaN,69.8,171,1
13,2024-03-04 06:44:50,Truck_2,-26.1836,-30.3641,348,Delivered,27.3,71.6,Detour,30,211,2,NaN,96.4,115,0
17,2024-07-15 15:42:02,Truck_7,-8.7105,-17.8347,344,Delayed,21.4,53.5,Detour,23,461,9,NaN,93.4,106,1
18,2024-01-19 06:42:30,Truck_5,14.6629,116.0696,409,Delayed,27.5,66.6,Detour,29,492,3,NaN,72.6,204,1


In [72]:
pd.crosstab(
    data['Logistics_Delay_Reason'].isnull(),
    data['Logistics_Delay'],
    margins=True
)

Logistics_Delay,0,1,All
Logistics_Delay_Reason,,,
False,318,419,737
True,116,147,263
All,434,566,1000


In [73]:
data['Logistics_Delay_Reason'].value_counts(dropna=False)


Logistics_Delay_Reason
Weather               267
NaN                   263
Traffic               236
Mechanical Failure    234
Name: count, dtype: int64

In [74]:
pd.crosstab(
    data['Logistics_Delay_Reason'],
    data['Logistics_Delay'],
    margins=True,
    dropna=False
)

Logistics_Delay,0,1,All
Logistics_Delay_Reason,,,
Mechanical Failure,101,133,234.0
Traffic,101,135,236.0
Weather,116,151,267.0
NaN,116,147,NaN
All,434,566,1000.0


In [24]:
# the above cross tab is suspiciously balanced 
# the Logistics_Delay_Reason may have been created independently of Logistics_Delay

'''
147 rows have a logistics delay even though Logistics_Delay_Reason is missing.

and

318 rows have a recorded delay reason even though Logistics_Delay = 0.
'''

'\n147 rows have a logistics delay even though Logistics_Delay_Reason is missing.\n\nand\n\n318 rows have a recorded delay reason even though Logistics_Delay = 0.\n'

In [ ]:
# fill blank values as "unknown"

In [75]:
data['Logistics_Delay_Reason'] = (
    data['Logistics_Delay_Reason'].fillna('Unknown')
)

In [76]:
data['Logistics_Delay_Reason'].value_counts()

Logistics_Delay_Reason
Weather               267
Unknown               263
Traffic               236
Mechanical Failure    234
Name: count, dtype: int64

In [77]:
# Check  duplicate rows
data.duplicated().sum()

0

In [47]:
data.shape

(1000, 16)

# Check categorical consistency

In [51]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Timestamp                1000 non-null   object 
 1   Asset_ID                 1000 non-null   object 
 2   Latitude                 1000 non-null   float64
 3   Longitude                1000 non-null   float64
 4   Inventory_Level          1000 non-null   int64  
 5   Shipment_Status          1000 non-null   object 
 6   Temperature              1000 non-null   float64
 7   Humidity                 1000 non-null   float64
 8   Traffic_Status           1000 non-null   object 
 9   Waiting_Time             1000 non-null   int64  
 10  User_Transaction_Amount  1000 non-null   int64  
 11  User_Purchase_Frequency  1000 non-null   int64  
 12  Logistics_Delay_Reason   1000 non-null   object 
 13  Asset_Utilization        1000 non-null   float64
 14  Demand_Forecast          

In [52]:
categorical_columns = [
    'Asset_ID',
    'Shipment_Status',
    'Traffic_Status',
    'Logistics_Delay_Reason'
]

for column in categorical_columns:
    print(f"\n{column}")
    print(data[column].value_counts())


Asset_ID
Asset_ID
Truck_8     109
Truck_4     107
Truck_10    105
Truck_2     105
Truck_6     103
Truck_7     102
Truck_9      94
Truck_3      93
Truck_5      93
Truck_1      89
Name: count, dtype: int64

Shipment_Status
Shipment_Status
Delayed       350
Delivered     338
In Transit    312
Name: count, dtype: int64

Traffic_Status
Traffic_Status
Detour    345
Clear     328
Heavy     327
Name: count, dtype: int64

Logistics_Delay_Reason
Logistics_Delay_Reason
Weather               267
Unknown               263
Traffic               236
Mechanical Failure    234
Name: count, dtype: int64


In [53]:
# check target data


data['Logistics_Delay'].value_counts()

Logistics_Delay
1    566
0    434
Name: count, dtype: int64

In [55]:

data['Logistics_Delay'].value_counts(normalize=True) * 100

Logistics_Delay
1    56.6
0    43.4
Name: proportion, dtype: float64

In [78]:
# No variations in how the same value was entered 

# Timestamp Column
- change from object to datatime

In [80]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype  
---  ------                   --------------  -----  
 0   Timestamp                1000 non-null   object 
 1   Asset_ID                 1000 non-null   object 
 2   Latitude                 1000 non-null   float64
 3   Longitude                1000 non-null   float64
 4   Inventory_Level          1000 non-null   int64  
 5   Shipment_Status          1000 non-null   object 
 6   Temperature              1000 non-null   float64
 7   Humidity                 1000 non-null   float64
 8   Traffic_Status           1000 non-null   object 
 9   Waiting_Time             1000 non-null   int64  
 10  User_Transaction_Amount  1000 non-null   int64  
 11  User_Purchase_Frequency  1000 non-null   int64  
 12  Logistics_Delay_Reason   1000 non-null   object 
 13  Asset_Utilization        1000 non-null   float64
 14  Demand_Forecast          

In [81]:
data.head()

,Timestamp,Asset_ID,Latitude,Longitude,Inventory_Level,Shipment_Status,Temperature,Humidity,Traffic_Status,Waiting_Time,User_Transaction_Amount,User_Purchase_Frequency,Logistics_Delay_Reason,Asset_Utilization,Demand_Forecast,Logistics_Delay
0,2024-03-20 00:11:14,Truck_7,-65.7383,11.2497,390,Delayed,27.0,67.8,Detour,38,320,4,Unknown,60.1,285,1
1,2024-10-30 07:53:51,Truck_6,22.2748,-131.7086,491,In Transit,22.5,54.3,Heavy,16,439,7,Weather,80.9,174,1
2,2024-07-29 18:42:48,Truck_10,54.9232,79.5455,190,In Transit,25.2,62.2,Detour,34,355,3,Unknown,99.2,260,0
3,2024-10-28 00:50:54,Truck_9,42.3900,-1.4788,330,Delivered,25.4,52.3,Heavy,37,227,5,Traffic,97.4,160,1
4,2024-09-27 15:52:58,Truck_7,-65.8477,47.9468,480,Delayed,20.5,57.2,Clear,56,197,6,Unknown,71.6,270,1


In [82]:
# Test timestamp conversion
timestamp_test = pd.to_datetime(
    data['Timestamp'],
    errors='coerce'
)

timestamp_test.isna().sum()

0

In [84]:
# change data type for Timestamp -> object to datetime

data['Timestamp'] = pd.to_datetime(data['Timestamp'])

In [85]:
data.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1000 entries, 0 to 999
Data columns (total 16 columns):
 #   Column                   Non-Null Count  Dtype         
---  ------                   --------------  -----         
 0   Timestamp                1000 non-null   datetime64[ns]
 1   Asset_ID                 1000 non-null   object        
 2   Latitude                 1000 non-null   float64       
 3   Longitude                1000 non-null   float64       
 4   Inventory_Level          1000 non-null   int64         
 5   Shipment_Status          1000 non-null   object        
 6   Temperature              1000 non-null   float64       
 7   Humidity                 1000 non-null   float64       
 8   Traffic_Status           1000 non-null   object        
 9   Waiting_Time             1000 non-null   int64         
 10  User_Transaction_Amount  1000 non-null   int64         
 11  User_Purchase_Frequency  1000 non-null   int64         
 12  Logistics_Delay_Reason   1000 non-n

In [86]:
# Check difference
print("Earliest timestamp:", data['Timestamp'].min())
print("Latest timestamp:", data['Timestamp'].max())

Earliest timestamp: 2024-01-01 11:37:57
Latest timestamp: 2024-12-30 20:21:58


In [87]:
# check for duplicates
data['Timestamp'].duplicated().sum()

0

In [88]:
''' Dataset is from Jan 1 2024 to Dec 30 2024 '''
# 2024 dataset

' Dataset is from Jan 1 2024 to Dec 30 2024 '

# Validate numerical variables

In [89]:
# Select numerical columns
numeric_columns = data.select_dtypes(include=['int64', 'float64']).columns

numeric_columns

Index(['Latitude', 'Longitude', 'Inventory_Level', 'Temperature', 'Humidity',
       'Waiting_Time', 'User_Transaction_Amount', 'User_Purchase_Frequency',
       'Asset_Utilization', 'Demand_Forecast', 'Logistics_Delay'],
      dtype='object')

In [90]:
# looking for suspicious or impossible values

# Check minimum and maximum values
range_check = pd.DataFrame({
    'Minimum': data[numeric_columns].min(),
    'Maximum': data[numeric_columns].max(),
    'Mean': data[numeric_columns].mean()
})

range_check

,Minimum,Maximum,Mean
Latitude,-89.7915,89.8701,-1.360093
Longitude,-179.8202,179.9237,0.837049
Inventory_Level,100.0000,500.0000,297.915000
Temperature,18.0000,30.0000,23.893900
Humidity,50.0000,80.0000,65.042200
Waiting_Time,10.0000,60.0000,35.062000
User_Transaction_Amount,100.0000,500.0000,299.055000
User_Purchase_Frequency,1.0000,10.0000,5.513000
Asset_Utilization,60.0000,100.0000,79.599100
Demand_Forecast,100.0000,300.0000,199.284000


- Latitude is ranging from the entire globe... 

Invalid latitude: 0
Invalid longitude: 0
Invalid humidity: 0
Invalid asset utilization: 0
Negative waiting time: 0
Negative inventory: 0
Negative demand forecast: 0
Invalid delay target: 0


# Logical consistency checks

# Shipment status vs. logistics delay


In [92]:
# Shipment status vs. logistics delay

shipment_delay = pd.crosstab(
    data['Shipment_Status'],
    data['Logistics_Delay'],
    margins=True
)

shipment_delay

Logistics_Delay,0,1,All
Shipment_Status,,,
Delayed,0,350,350
Delivered,220,118,338
In Transit,214,98,312
All,434,566,1000


In [93]:
# percentages 

pd.crosstab(
    data['Shipment_Status'],
    data['Logistics_Delay'],
    normalize='index'
).round(3) * 100

Logistics_Delay,0,1
Shipment_Status,,
Delayed,0.0,100.0
Delivered,65.1,34.9
In Transit,68.6,31.4


# Traffic conditions vs logistics delay


In [94]:
# Traffic conditions vs logistics delay

traffic_delay = pd.crosstab(
    data['Traffic_Status'],
    data['Logistics_Delay'],
    margins=True
)

traffic_delay

Logistics_Delay,0,1,All
Traffic_Status,,,
Clear,213,115,328
Detour,221,124,345
Heavy,0,327,327
All,434,566,1000


In [95]:
pd.crosstab(
    data['Traffic_Status'],
    data['Logistics_Delay'],
    normalize='index'
).round(3) * 100

Logistics_Delay,0,1
Traffic_Status,,
Clear,64.9,35.1
Detour,64.1,35.9
Heavy,0.0,100.0


- Delay reason vs. logistics delay
- Would a company know the reason for a delay before predicting whether the shipment will be delayed?
- Most likely No

In [96]:
# Inventory vs forecast

# Check whether forecast demand exceeds available inventory
(data['Demand_Forecast'] > data['Inventory_Level']).value_counts()

False    739
True     261
Name: count, dtype: int64

In [97]:
(
    (data['Demand_Forecast'] > data['Inventory_Level'])
    .value_counts(normalize=True) * 100
).round(2)

False    73.9
True     26.1
Name: proportion, dtype: float64

More than one-quarter of observations have forecast demand exceeding available inventory.

# Identify Outliers
- IQR METHOD

In [99]:
'''
Remove columns that are numeric but finding their outliers will not be beneficial

- Latitude
- Longitude
- Logistics_Delay  (binary)
'''

# Numerical columns to check for potential outliers
outlier_columns = [
    'Inventory_Level',
    'Temperature',
    'Humidity',
    'Waiting_Time',
    'User_Transaction_Amount',
    'User_Purchase_Frequency',
    'Asset_Utilization',
    'Demand_Forecast'
]

outlier_summary = []

for column in outlier_columns:
    
    Q1 = data[column].quantile(0.25)
    Q3 = data[column].quantile(0.75)
    
    IQR = Q3 - Q1
    
    lower_bound = Q1 - 1.5 * IQR
    upper_bound = Q3 + 1.5 * IQR
    
    outliers = data[
        (data[column] < lower_bound) |
        (data[column] > upper_bound)
    ]
    
    outlier_summary.append({
        'Variable': column,
        'Lower_Bound': round(lower_bound, 2),
        'Upper_Bound': round(upper_bound, 2),
        'Outlier_Count': len(outliers)
    })

pd.DataFrame(outlier_summary)

,Variable,Lower_Bound,Upper_Bound,Outlier_Count
0,Inventory_Level,-96.00,696.00,0
1,Temperature,13.10,34.70,0
2,Humidity,34.40,95.20,0
3,Waiting_Time,-16.00,88.00,0
4,User_Transaction_Amount,-128.12,724.88,0
5,User_Purchase_Frequency,-4.50,15.50,0
6,Asset_Utilization,39.55,119.35,0
7,Demand_Forecast,-16.88,412.12,0


In [100]:
# NO OUTLIERS DETECTED WITH IQR METHOD

In [101]:
# Final data quality check
print("Dataset shape:", data.shape)
print("Missing values:", data.isnull().sum().sum())
print("Duplicate rows:", data.duplicated().sum())
print("\nData types:")
print(data.dtypes)

Dataset shape: (1000, 16)
Missing values: 0
Duplicate rows: 0

Data types:
Timestamp                  datetime64[ns]
Asset_ID                           object
Latitude                          float64
Longitude                         float64
Inventory_Level                     int64
Shipment_Status                    object
Temperature                       float64
Humidity                          float64
Traffic_Status                     object
Waiting_Time                        int64
User_Transaction_Amount             int64
User_Purchase_Frequency             int64
Logistics_Delay_Reason             object
Asset_Utilization                 float64
Demand_Forecast                     int64
Logistics_Delay                     int64
dtype: object


In [4]:
# Save the cleaned dataset

data.to_csv(
    '../data/smart_logistics_cleaned.csv',
    index=False
)